### ISSUE TITLE
Hamming distance returns unexpected value for string comparisons

### ISSUE THREAD

**Issue by fabianegli** [scipy/scipy#11991]
**Problem statement and current scipy behavior**

The [scipy.spatial.distance.hamming](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.hamming.html) distance metric returns 1.0 for unequal strings but returns the expected values if strings are converted to lists beforehand. 

The following code shows the output generated by the current version of scipy. Note the difference between the distance which is expected to be 0.25 for the chosen strings.

```Python
from scipy.spatial import distance
>>> distance.hamming("same", "tame")
1.0
>>> distance.hamming(list("same"), list("tame"))
0.25
```

**Desired behavior**

```Python
from scipy.spatial import distance
>>> distance.hamming("same", "tame")
0.25
```

**Comment by fabianegli** [comment]
Upon further testing, hamming also returns 1.0 for strings of unequal lengths, which should raise a ValueError like the one when arrays of unequal lengths are supplied.

```Python
distance.hamming("hello", "hell")
1.0
distance.hamming([1,2], [1])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-17-29e1919c2024> in <module>
----> 1 distance.hamming([1,2], [1])

.../scipy/spatial/distance.py in hamming(u, v, w)
    814     v = _validate_vector(v)
    815     if u.shape != v.shape:
--> 816         raise ValueError('The 1d arrays must have equal lengths.')
    817     u_ne_v = u != v
    818     if w is not None:

ValueError: The 1d arrays must have equal lengths.
```

**Comment by WarrenWeckesser** [comment]
This is to be expected.  The first thing the distance functions do is convert their arguments to numpy arrays.  When converted to a numpy array, a single string is treated as a scalar, not as a sequence of characters.  E.g.
```
In [5]: np.array("same")
Out[5]: array('same', dtype='<U4')
```
vs.
```
In [6]: np.array(list("same"))                                                  
Out[6]: array(['s', 'a', 'm', 'e'], dtype='<U1')
```

**Comment by fabianegli** [comment]
Yes, you are of course correct with the description of the current behavior. I just expect a different behavior of the `hamming` function when strings are supplied. 

By the docstring this function `Compute[s] the Hamming distance between two 1-D arrays.`. This means it should basically raise a ValueError for any non-array input. But it makes the assumption that the direct conversion to a numpy array will result in the desired conversion if something else is supplied. I argue that this is not the case and my fundamental assumption is that in the case when both arguments are strings the user desires a character by character comparison. To get that the strings need to be split before the numpy array conversion.

**Comment by fabianegli** [comment]
PR #12027 also contains some more reasoning for the proposed change. 

Does the proposed change violate some design choices scipy made? Or would it be more desirable to create a separate hamming function for strings?

**Comment by h-vetinari** [comment]
I think this might be the problem regarding the normalization mentioned [here](https://github.com/scipy/scipy/pull/3163#issuecomment-34224341):

> 1. `hamming` doesn't compute the Hamming distance, but scales it dividing by the vector dimension. They are closely related, but are not quite the same. I would like to have `matching` and `sokalmichener` return the old value, but make `hamming` return the proper unscaled Hamming distance. Can this be happily done, or is it best to first add some form of deprecation warning?

**Comment by fabianegli** [comment]
@h-vetinari This issue does not touch the issue of normalization. It is purely about the handling of arguments that are strings - even if this issue/PR is merged, the hamming function will still calculate the normalized hamming distance.

**Comment by h-vetinari** [comment]
> @h-vetinari This issue does not touch the issue of normalization.

Your OP example from the OP

> ```python
> from scipy.spatial import distance
> >>> distance.hamming("same", "tame")
> 1.0
> >>> distance.hamming(list("same"), list("tame"))
> 0.25
> ```

looks like there's implicit normalization by the length of the list, which is why I commented. You might _want_ that normalization, but it's a point where reasonable people can disagree. I think you suggested in another issue to add a `normalize`-kwarg, which sounds useful to serve both camps.

PS. No normalization is what I'd intuitively expect from what exposure I had to hamming-distances during my undergrad.

**Comment by fabianegli** [comment]
@h-vetinari Yes, hamming does normalize. This is, however, standard behaviour and not something this issue touches. Read: this issue is about the interpretation of the function arguments, not how to calculate distance metric. At the heart of this issue is the question if string arguments should be handled as numpy scalars or as list of characters. Please do not further discuss the normalization here. If you would like to discuss the normalization, you should do that in a new issue.

**Comment by h-vetinari** [comment]
Clearly, strings are currently being handled as scalars:
```python
>>> distance.hamming("same", "tamer")
1.0
>>> distance.hamming(list("same"), list("tamer"))
ValueError: The 1d arrays must have equal lengths.
>>> distance.hamming("same", "tale")
1.0  # not 2.0
>>> distance.hamming(list("same"), list("tale"))
0.5
```

What we see is that the hamming-distance between scalar strings `x` & `y` is equivalent to `x!=y` (and the dimension therefore pretty degenerate, but consistent with the strings being scalar).

You now propose that strings be interpreted as a list of characters - why? How is this behaviour more consistent (and/or more explicit) than the current one? I'd be _more_ surprised if a string & a list of characters yielded the same result.

**Comment by fabianegli** [comment]
@h-vetinari please read through the PR #12027 to better understand the reasoning for the proposed changes and why it is an issue. You will also read that I now tend to think that hamming should raise a ValueError if the arguments are not array like.

To simply answer your question about why to interpret a string as a list of characters: they are. And Python in some context explicitly handles them as such, e.g. loop through a string, slicing a string and extracting character the same way as you do with list elements. That is all standard Python. So `for character in "string": print(character)` and `for character in list("string"): print(character)` is the equivalent.

**Comment by fabianegli** [comment]
Also, consider the code example for the hamming distance from [Wikipedia](https://en.wikipedia.org/wiki/Hamming_distance). 

```python
def hamming_distance(s1, s2) -> int:
    """Return the Hamming distance between equal-length sequences."""
    if len(s1) != len(s2):
        raise ValueError("Undefined for sequences of unequal length.")
    return sum(el1 != el2 for el1, el2 in zip(s1, s2))
```

I don't think it is an authoritative source in this respect, but still conveys that strings are sometimes interpreted as sequences, lists or array (all these names convey the same basic concept).

Python has a list like interpretation of stings while numpy sees them as scalars. In scipy the numpy and Python world collide with their different interpretation of strings, hence the scipy hamming function has to either chose one of the two views of what a string is or not choose at all and reject strings to force the user to be specific and make that choice consciously.

**Comment by ilayn** [comment]
Probably we should get rid of the naked string inputs. since they are not really array-like and text distance tools are ubiquitous everywhere else 

```python
[x == y for x, y in zip(a, b)].count(True)/len(a)
```

would probably better anyways for strings (didn't measure). Or if this particular version is needed inputs can be converted to arrays. 

But at the same time, it is a waste of overhead to test every input type so I'll be also OK if we document this and get on with it.

**Comment by fabianegli** [comment]
Just to recap the value proposition for the string comparison: Generally, the hamming distance does not apply to scalars, but lists/sequences/arrays. So if any of the arguments to hamming passes the `numpy.isscalar` test, raising a ValueError would be the logical thing to do.

Handling string input as scalars makes no sense, because as stated above the distance does not apply to scalars plus the current behaviour just emulates a simple equality comparison. Because strings inherently contain a sequence of letters, the interpretation as a list is possible and I argue sensical. Calculating the character wise hamming distance by default provides value because it doesn't require the explicit conversion into a list be done by the user.

The reason why hamming function now works the way it does is the fact that it internally uses `_validate_vector(argument)` which in first converts the argument with `numpy.asarray(argument)`. hamming then numpy operations on these arrays that might (or might not) be speedier than the element wise comparison proposed by @ilayn / Wikipedia (depending on the size and content of the lists). The array conversion leaves normal python lists in a comparable state but "wraps" scalars into a list - that's where the magic happens, or in this case the problem arises.

After looking at the distance.py file again I now think that the numpy.isscalar test should be done in the `_validate_vector` function before the np.asarray conversion takes place. But this would then affect all the distance metrics using the that input checking function.

Sidenote: zip consumes any iterable. Consider something like `list(zip("string", list("string")))` which makes it obvious that there is a big potential for unanticipated behaviour of that code.

I know input type checking is an overhead, but in this case it handles a special case and adds functionality which removes that overhead from the waste category.

**Comment by h-vetinari** [comment]
> So if any of the arguments to hamming passes the `numpy.isscalar` test, raising a ValueError would be the logical thing to do.

This I would agree with, ideally with a clear error message. Then the user can turn the string into an array, and it becomes a non-issue.

**Comment by fabianegli** [comment]
> This I would agree with, ideally with a clear error message. Then the user can turn the string into an array, and it becomes a non-issue.

If the newly proposed behaviour is indeed desired by, I am happy to create a new Issue+PR specifically for that. Reason being is that this change will not be hamming distance specific and is not the scope of this issue and corresponding PR. 

@ilayn and @WarrenWeckesser can you please give an emojy reaction thumbs-up or thumbs-down so I have an indication if the more stringent argument requirements would end up in scipy or not. thumbs-up would men "go for it" and thumbs-down "don't waste your time".
